supermarq/features.py

```
@@ -0,0 +1,163 @@
+
+    return degree_sum / (N * (N - 1))
+
+
+def compute_liveness(circuit: Union[cirq.Circuit, qiskit.circuit.QuantumCircuit]) -> float:
+    """
+    liveness feature = sum of all entries in the liveness matrix / (num_qubits * depth)
+
+    Input
+    -----
+    circ : A cirq or qiskit quantum circuit
+    """
+
+    if isinstance(circuit, cirq.Circuit):
+        circ = sm.converters.cirq_to_qiskit(circuit)
+    else:
+        circ = circuit
+
+    N = circ.num_qubits
+    dag = qiskit.converters.circuit_to_dag(circ)
+    dag.remove_all_ops_named("barrier")
+
+    activity_matrix = np.zeros((N, dag.depth()))
+
+    for i, layer in enumerate(dag.layers()):
+        for op in layer["partition"]:
+            for qubit in op:
+                activity_matrix[qubit.index, i] = 1
+
+    return np.sum(activity_matrix) / (N * dag.depth())
+
+
+def compute_parallelism(circuit: Union[cirq.Circuit, qiskit.circuit.QuantumCircuit]) -> float:
+    """
+    parallelism feature = max(1 - depth / # of gates, 0)
+
+    Input
+    -----
+    circ : A cirq or qiskit quantum circuit
+    """
+
+    if isinstance(circuit, cirq.Circuit):
+        circ = sm.converters.cirq_to_qiskit(circuit)
+    else:
+        circ = circuit
+
+    dag = qiskit.converters.circuit_to_dag(circ)
+    dag.remove_all_ops_named("barrier")
+    return max(1 - (circ.depth() / len(dag.gate_nodes())), 0)
+
+
+def compute_measurement(circuit: Union[cirq.Circuit, qiskit.circuit.QuantumCircuit]) -> float:
+    """
+    measurement feature = # of layers of mid-circuit measurement / circuit depth
+
+    Input
+    -----
+    circ : A cirq or qiskit quantum circuit
+    """
+    if isinstance(circuit, cirq.Circuit):
+        circ = sm.converters.cirq_to_qiskit(circuit)
+    else:
+        circ = circuit
+
+    dag = qiskit.converters.circuit_to_dag(circ)
+    dag.remove_all_ops_named("barrier")
+
+    reset_moments = 0
+    gate_depth = dag.depth() - 1  # assumes that circuit measures all qubits at the end
```

Does qiskit have a way to assert that measurements are indeed terminal before we call this? Seems like currently there is a way for this to silently fail.

---

Good catch -- updated this so instead of potentially dividing by zero, we will remove the final measurements in the circuit if they exist.